formula/src/test/java/com/instacart/formula/FormulaRuntimeTest.kt

```diff
@@ -212,6 +214,113 @@ class FormulaRuntimeTest(val runtime: TestableRuntime, val name: String) {
         robot.test.output { assertThat(this).isEqualTo(3) }
     }
 
+    @Test fun `input change while running triggers root formula restart`() {
+        val terminationCallback = TestEventCallback<Int>()
+        var observer: TestFormulaObserver<Int, *, *>? = null
+        val root = object : StatelessFormula<Int, Int>() {
+            override fun key(input: Int): Any {
+                return input
+            }
+
+            override fun Snapshot<Int, Unit>.evaluate(): Evaluation<Int> {
+                return Evaluation(
+                    output = input,
+                    actions = context.actions {
+                        Action.onTerminate().onEvent {
+                            transition {
+                                terminationCallback.invoke(input)
+                            }
+                        }
+
+                        if (input == 0) {
+                            Action.onInit().onEvent {
+                                /**
+                                 * We call observer explicitly outside of effect block to ensure
+                                 * that input change happens while formula is running
+                                 */
+                                observer?.input(1)
+                                none()
+                            }
+                        }
+                    }
+                )
+            }
+        }
+
+        observer = runtime.test(root)
+        observer.input(0)
+        observer.output { assertThat(this).isEqualTo(1) }
+
+        // Check that termination was called
+        assertThat(terminationCallback.values()).containsExactly(0).inOrder()
+    }
+
+    @Test fun `runtime termination triggered while formula is running`() {
+        val terminationCallback = TestEventCallback<Int>()
+        var observer: TestFormulaObserver<Int, *, *>? = null
+        val root = object : StatelessFormula<Int, Int>() {
+            override fun Snapshot<Int, Unit>.evaluate(): Evaluation<Int> {
+                return Evaluation(
+                    output = input,
+                    actions = context.actions {
+                        Action.onTerminate().onEvent {
+                            transition {
+                                terminationCallback.invoke(input)
+                            }
+                        }
+
+                        if (input == 0) {
+                            Action.onInit().onEvent {
+                                // This is outside of effect to trigger termination while running
+                                observer?.dispose()
+                                none()
+                            }
+                        }
+                    }
+                )
+            }
+        }
+
+        observer = runtime.test(root)
+        observer.input(0)
```

Comment:
This wouldn't have a valid output after terminating, right?

Reply:
Good point, will validate